File: src/CSharpCore/DreamEngine/Sync/EntitySynchronizer.cs
using System.Globalization;
using System.Reflection;
using DreamEngine.Core;
using DreamEngine.ECS;
using DreamEngine.ECS.Components;
using DreamEngine.Extensions;
using DreamEngine.Sync.Data;

namespace DreamEngine.Sync;

internal static class EntitySynchronizer
{
    // must use ref here, because we want to modify the entityDTO
    // without ref, c# pass it a copy of that struct
    public static void SynchronizeFromTo(Entity entity, ref EntityData entityData)
    {
        entityData.active = entity.Active.ToByte();
        entityData.parentId = entity.Parent?.Id ?? 1;
        entityData.parentHas = (byte)(entity.Parent is not null ? 1 : 0);

        entityData.transformPositionX = entity.Transform.Position.X;
        entityData.transformPositionY = entity.Transform.Position.Y;
        entityData.transformPositionZ = entity.Transform.Position.Z;
        entityData.transformRotationX = entity.Transform.Rotation.X;
        entityData.transformRotationY = entity.Transform.Rotation.Y;
        entityData.transformRotationZ = entity.Transform.Rotation.Z;
        entityData.transformScaleX = entity.Transform.Scale.X;
        entityData.transformScaleY = entity.Transform.Scale.Y;
        entityData.transformScaleZ = entity.Transform.Scale.Z;

        // ### Components to EntityData ###

        // ui component
        entityData.uiComponentHas = HasComponentInByte(entity.GetComponent<UiComponent>());

        // camera component
        var cameraComponent = entity.GetComponent<CameraComponent>();
        entityData.cameraComponentHas = HasComponentInByte(cameraComponent);
        
        if (cameraComponent is not null)
        {
            entityData.cameraFar = cameraComponent.Far;
            entityData.cameraNear = cameraComponent.Near;
            entityData.cameraFov = cameraComponent.Fov;
        }

        // directional light component
        var directionalLightComponent = entity.GetComponent<DirectionalLightComponent>();
        entityData.directionalLightComponentHas = HasComponentInByte(directionalLightComponent);

        if (directionalLightComponent is not null)
        {
            entityData.directionalLightColorRed = directionalLightComponent.Color.Red;
            entityData.directionalLightColorGreen = directionalLightComponent.Color.Green;
            entityData.directionalLightColorBlue = directionalLightComponent.Color.Blue;
            entityData.directionalLightColorAlpha = directionalLightComponent.Color.Alpha;

            entityData.directionalLightSpecularX = directionalLightComponent.Specular.X;
            entityData.directionalLightSpecularY = directionalLightComponent.Specular.Y;
            entityData.directionalLightSpecularZ = directionalLightComponent.Specular.Z;

            entityData.directionalLightInfluence = directionalLightComponent.Influence;
        }

        // collider component
        var colliderComponent = entity.GetComponent<ColliderComponent>();
        entityData.colliderComponentHas = HasComponentInByte(colliderComponent);

        if (colliderComponent is not null)
        {
            entityData.colliderEnabled = colliderComponent.Enabled.ToByte();
            entityData.colliderIsTrigger = colliderComponent.IsTrigger.ToByte();

            entityData.colliderCenterX = colliderComponent.Center.X;
            entityData.colliderCenterY = colliderComponent.Center.Y;
            entityData.colliderCenterZ = colliderComponent.Center.Z;

            entityData.colliderSizeX = colliderComponent.Size.X;
            entityData.colliderSizeY = colliderComponent.Size.Y;
            entityData.colliderSizeZ = colliderComponent.Size.Z;
        }

        // audio listener component
        var audioListenerComponent = entity.GetComponent<AudioListenerComponent>();
        entityData.audioListenerComponentHas = HasComponentInByte(audioListenerComponent);

        if (audioListenerComponent is not null)
            entityData.audioListenerEnabled = audioListenerComponent.Enabled.ToByte();

        // audio emitter component
        var audioEmitterComponent = entity.GetComponent<AudioEmitterComponent>();
        entityData.audioEmitterComponentHas = HasComponentInByte(audioEmitterComponent);

        if (audioEmitterComponent is not null)
        {
            entityData.audioEmitterEnabled = audioEmitterComponent.Enabled.ToByte();
            entityData.audioEmitterSpatial = audioEmitterComponent.Spatial.ToByte();

            entityData.audioEmitterMaxDistance = audioEmitterComponent.MaxDistance;
            entityData.audioEmitterMinDistance = audioEmitterComponent.MinDistance;
            entityData.audioEmitterPitch = audioEmitterComponent.Pitch;
            entityData.audioEmitterVolume = audioEmitterComponent.Volume;
        }
    }

    // using the ref keyword is the same case as above
    public static void SynchronizeFromTo(ref EntityData entityData, Entity entity)
    {
        entity.Id = entityData.id;
        entity.Active = entityData.active == 1;

        if (entityData.parentHas == 1)
            entity.Parent = Game.Scene.GetEntity(entityData.parentId);

        entity.Transform.Position.InternalX = entityData.transformPositionX;
        entity.Transform.Position.InternalY = entityData.transformPositionY;
        entity.Transform.Position.InternalZ = entityData.transformPositionZ;
        entity.Transform.Rotation.InternalX = entityData.transformRotationX;
        entity.Transform.Rotation.InternalY = entityData.transformRotationY;
        entity.Transform.Rotation.InternalZ = entityData.transformRotationZ;
        entity.Transform.Scale.InternalX = entityData.transformScaleX;
        entity.Transform.Scale.InternalY = entityData.transformScaleY;
        entity.Transform.Scale.InternalZ = entityData.transformScaleZ;

        HandleComponentToEntity<UiComponent>(ref entityData, entity, entityData.uiComponentHas);
        HandleComponentToEntity<CameraComponent>(ref entityData, entity, entityData.cameraComponentHas, HandleCreateCameraComponent);
        HandleComponentToEntity<DirectionalLightComponent>(ref entityData, entity, entityData.directionalLightComponentHas, HandleCreateDirectionalLightComponent);
        HandleComponentToEntity<ColliderComponent>(ref entityData, entity, entityData.colliderComponentHas, HandleCreateColliderComponent);
        HandleComponentToEntity<AudioListenerComponent>(ref entityData, entity, entityData.audioListenerComponentHas, HandleCreateAudioListenerComponent);
        HandleComponentToEntity<AudioEmitterComponent>(ref entityData, entity, entityData.audioEmitterComponentHas, HandleCreateAudioEmitterComponent);
    }

    private static byte HasComponentInByte(Component? component)
    {
        return (byte)(component is not null ? 1 : 0);
    }

    private static void HandleComponentToEntity<T>(ref EntityData entityData, Entity entity, byte hasComponent, Action<EntityData, T>? action = null) 
        where T : Component
    {
        var component = entity.GetComponent<T>();

        if (hasComponent == 0)
        {
            if (component is null)
                return;

            entity.RemoveComponent(component);

            return;
        }

        component ??= entity.AddComponent<T>();
        action?.Invoke(entityData, component);
    }

    private static void HandleCreateCameraComponent(EntityData entityData, CameraComponent cameraComponent)
    {
        cameraComponent.Far = entityData.cameraFar;
        cameraComponent.Fov = entityData.cameraFov;
        cameraComponent.Near = entityData.cameraNear;
    }

    private static void HandleCreateDirectionalLightComponent(EntityData entityData, DirectionalLightComponent directionalLightComponent)
    {
        directionalLightComponent.Color.Red = entityData.directionalLightColorRed;
        directionalLightComponent.Color.Green = entityData.directionalLightColorGreen;
        directionalLightComponent.Color.Blue = entityData.directionalLightColorBlue;
        directionalLightComponent.Color.Alpha = entityData.directionalLightColorAlpha;

        directionalLightComponent.Specular.X = entityData.directionalLightSpecularX;
        directionalLightComponent.Specular.Y = entityData.directionalLightSpecularY;
        directionalLightComponent.Specular.Z = entityData.directionalLightSpecularZ;

        directionalLightComponent.Influence = entityData.directionalLightInfluence;
    }

    private static void HandleCreateColliderComponent(EntityData entityData, ColliderComponent colliderComponent)
    {
        colliderComponent.Enabled = entityData.colliderEnabled == 1;
        colliderComponent.IsTrigger = entityData.colliderIsTrigger == 1;

        colliderComponent.Center.X = entityData.colliderCenterX;
        colliderComponent.Center.Y = entityData.colliderCenterY;
        colliderComponent.Center.Z = entityData.colliderCenterZ;

        colliderComponent.Size.X = entityData.colliderSizeX;
        colliderComponent.Size.Y = entityData.colliderSizeY;
        colliderComponent.Size.Z = entityData.colliderSizeZ;
    }

    private static void HandleCreateAudioListenerComponent(EntityData data, AudioListenerComponent component)
    {
        component.Enabled = data.audioListenerEnabled == 1;
    }

    private static void HandleCreateAudioEmitterComponent(EntityData data, AudioEmitterComponent component)
    {
        component.Enabled = data.audioEmitterEnabled == 1;
        component.Spatial = data.audioEmitterSpatial == 1;

        component.MaxDistance = data.audioEmitterMaxDistance;
        component.MinDistance = data.audioEmitterMinDistance;
        component.Pitch = data.audioEmitterPitch;
        component.Volume = data.audioEmitterVolume;
    }
}

File: src/Core/Sync/EntitySynchronizer.cpp
#include "EntitySynchronizer.h"
#include "../ECS/Components/TransformComponent.h"
#include "ECS/Components/ScriptComponent.h"
#include "ECS/Components/UiComponent.h"
#include "ECS/Components/DirectionalLightComponent.h"
#include "ECS/Components/CameraComponent.h"
#include "ECS/Components/ChildrenComponent.h"
#include "Application.h"
#include "glm/gtx/quaternion.hpp"

using namespace DreamEngine::Core::Sync;
using namespace DreamEngine::Core::ECS::Components;

void EntitySynchronizer::SynchronizeToData(Entity* entity)
{
    TransformComponent& transformComponent = entity->GetComponent<TransformComponent>();
    const glm::vec3 position = transformComponent.GetPosition();
    const glm::vec3 rotation = transformComponent.GetRotation();
    const glm::vec3 scale = transformComponent.GetScale();

    entity->entityData.id = entity->GetId();
    entity->entityData.transformPositionX = position.x;
    entity->entityData.transformPositionY = position.y;
    entity->entityData.transformPositionZ = position.z;
    entity->entityData.transformRotationX = rotation.x;
    entity->entityData.transformRotationY = rotation.y;
    entity->entityData.transformRotationZ = rotation.z;
    entity->entityData.transformScaleX = scale.x;
    entity->entityData.transformScaleY = scale.y;
    entity->entityData.transformScaleZ = scale.z;

    // parent component
    ParentComponent& parentComponent = entity->GetComponent<ParentComponent>();

    if (parentComponent.has && parentComponent.parent != nullptr)
        entity->entityData.parentId = parentComponent.parent->GetId();

    // ui component
    UiComponent& uiComponent = entity->GetComponent<UiComponent>();
    entity->entityData.uiComponentHas = uiComponent.has;

    // script info
    ScriptComponent& scriptComponent = entity->GetComponent<ScriptComponent>();
    if (scriptComponent.has && scriptComponent.script != nullptr)
    {
        std::string assembly = scriptComponent.script->GetAssemblyName();
        std::strncpy(entity->entityData.scriptAssemblyName, assembly.c_str(), sizeof(entity->entityData.scriptAssemblyName) - 1);
        entity->entityData.scriptAssemblyName[sizeof(entity->entityData.scriptAssemblyName) - 1] = '\0';

        std::string className = scriptComponent.script->GetClassName();
        std::strncpy(entity->entityData.scriptClassName, className.c_str(), sizeof(entity->entityData.scriptClassName) - 1);
        entity->entityData.scriptClassName[sizeof(entity->entityData.scriptClassName) - 1] = '\0';
    }

    // directional light component
    DirectionalLightComponent& directionalLightComponent = entity->GetComponent<DirectionalLightComponent>();
    entity->entityData.directionalLightComponentHas = directionalLightComponent.has;

    if (directionalLightComponent.has)
    {
        entity->entityData.directionalLightSpecularX = directionalLightComponent.specular.x;
        entity->entityData.directionalLightSpecularY = directionalLightComponent.specular.y;
        entity->entityData.directionalLightSpecularZ = directionalLightComponent.specular.z;

        entity->entityData.directionalLightInfluence = directionalLightComponent.influence;

        entity->entityData.directionalLightColorRed = directionalLightComponent.color.red;
        entity->entityData.directionalLightColorGreen = directionalLightComponent.color.green;
        entity->entityData.directionalLightColorBlue = directionalLightComponent.color.blue;
        entity->entityData.directionalLightColorAlpha = directionalLightComponent.color.alpha;
    }

    // camera component
    CameraComponent& cameraComponent = entity->GetComponent<CameraComponent>();
    entity->entityData.cameraComponentHas = cameraComponent.has;

    if (entity->entityData.cameraComponentHas)
    {
        entity->entityData.cameraFar = cameraComponent.far;
        entity->entityData.cameraFov = cameraComponent.fovDegree;
        entity->entityData.cameraNear = cameraComponent.near;
    }
}

void EntitySynchronizer::SynchronizeFromData(Entity* entity)
{
    TransformComponent& transformComponent = entity->GetComponent<TransformComponent>();

    // parent id component
    HandleParentComponent(entity);

    // Extract individual components from EntityData
    glm::vec3 scale(entity->entityData.transformScaleX, entity->entityData.transformScaleY, entity->entityData.transformScaleZ);
    glm::vec3 rotationEuler(glm::radians(entity->entityData.transformRotationX), 
                            glm::radians(entity->entityData.transformRotationY),
                            glm::radians(entity->entityData.transformRotationZ));
    glm::vec3 position(entity->entityData.transformPositionX, entity->entityData.transformPositionY, entity->entityData.transformPositionZ);

    // Update the TransformComponent's individual properties
    transformComponent.SetScale(scale);
    transformComponent.SetRotation(rotationEuler);
    transformComponent.SetPosition(position);

    // Construct the transformation matrix: Scale -> Rotate -> Translate
    glm::mat4 scaleMatrix = glm::scale(glm::mat4(1.0f), scale);
    glm::mat4 rotationMatrix = glm::toMat4(glm::quat(rotationEuler));
    glm::mat4 translationMatrix = glm::translate(glm::mat4(1.0f), position);

    glm::mat4 transformMatrix = translationMatrix * rotationMatrix * scaleMatrix;

    // Update the TransformComponent's transform matrix
    transformComponent.SetTransform(transformMatrix);

    // update ui component
    UiComponent& uiComponent = entity->GetComponent<UiComponent>();
    uiComponent.has = entity->entityData.uiComponentHas;

    // update directional light component
    DirectionalLightComponent& directionalLightComponent = entity->GetComponent<DirectionalLightComponent>();
    directionalLightComponent.has = entity->entityData.directionalLightComponentHas;
    directionalLightComponent.color.red = entity->entityData.directionalLightColorRed;
    directionalLightComponent.color.green = entity->entityData.directionalLightColorGreen;
    directionalLightComponent.color.blue = entity->entityData.directionalLightColorBlue;
    directionalLightComponent.color.alpha = entity->entityData.directionalLightColorAlpha;

    directionalLightComponent.specular.x = entity->entityData.directionalLightSpecularX;
    directionalLightComponent.specular.y = entity->entityData.directionalLightSpecularY;
    directionalLightComponent.specular.z = entity->entityData.directionalLightSpecularZ;

    directionalLightComponent.influence = entity->entityData.directionalLightInfluence;

    // update camera component
    CameraComponent& cameraComponent = entity->GetComponent<CameraComponent>();
    cameraComponent.has = entity->entityData.cameraComponentHas;
    cameraComponent.far = entity->entityData.cameraFar;
    cameraComponent.fovDegree = entity->entityData.cameraFov;
    cameraComponent.near = entity->entityData.cameraNear;
}

void DreamEngine::Core::Sync::EntitySynchronizer::HandleParentComponent(Entity* entity)
{
    ParentComponent& parentComponent = entity->GetComponent<ParentComponent>();
    parentComponent.has = entity->entityData.parentId > 0;

    if (parentComponent.parent != nullptr && (parentComponent.parent->GetId() != entity->entityData.parentId || entity->entityData.parentId <= 0) ||
        parentComponent.parent == nullptr && entity->entityData.parentId > 0)
    {
        RemoveEntityFromParent(entity, parentComponent);
        AddEntityToParent(entity, parentComponent);
    }
}

void DreamEngine::Core::Sync::EntitySynchronizer::RemoveEntityFromParent(Entity* entity, ParentComponent& parentComponent)
{
    if (parentComponent.parent == nullptr)
        return;

    unsigned int entityId = entity->GetId();
    ChildrenComponent& oldParentChildrenComp = parentComponent.parent->GetComponent<ChildrenComponent>();

    std::vector<Entity*>::iterator childrenFound =
        std::find_if(oldParentChildrenComp.children.begin(), oldParentChildrenComp.children.end(), [entityId](Entity* ent) { return ent->GetId() == entityId; });

    if (childrenFound != oldParentChildrenComp.children.end())
    {
        oldParentChildrenComp.children.erase(childrenFound);
    }
}

void DreamEngine::Core::Sync::EntitySynchronizer::AddEntityToParent(Entity* entity, ParentComponent& parentComponent)
{
    if (entity->entityData.parentId <= 0)
    {
        parentComponent.parent = nullptr;

        return;
    }

    unsigned int parentId = entity->entityData.parentId;
    std::vector<Entity*> entities = Application::Instance().GetGame()->GetActiveScene()->GetEntityManager()->GetEntities();
    std::vector<Entity*>::iterator it = std::find_if(entities.begin(), entities.end(), [parentId](Entity* pEntity) { return pEntity->GetId() == parentId; });

    if (it != entities.end())
    {
        parentComponent.parent = (*it);
        ChildrenComponent& parentChildrenComponent = parentComponent.parent->GetComponent<ChildrenComponent>();
        parentChildrenComponent.children.push_back(entity);
    }
}
